### ISSUE TITLE
training_step(...) return type, without computational graph.

### ISSUE THREAD

**Issue by fperazzi** [PyTorchLightning/pytorch-lightning#3262]
Similar to the GAN example, assume that I have two networks G, D within the same `LightingModule`. Differently from it, the D network kicks in only after N epochs and I don't have a loss with a computational graph attached till then. What should the `training_step(...)` function return in this case? Is there a proper way to handle this case or should I return a dummy loss?

**Comment by awaelchli** [comment]
Yes, I think this is a good way, but then you have to override optimizer_step in order to prevent backprop the gradient from this dummy loss. 

I discussed once with @williamFalcon the possibility to allow returning None when using mutliple optimizers as long as one of the calls returns a loss, as in your case. When returning None, then this would also allow you to skip the forward pass completely instead of wasting it. I think this would simplify modeling gans a lot since we don't have to override optimizer step anymore. 

What do you think?

**Comment by fperazzi** [comment]
@awaelchli, sounds like a good idea to me. I had to to override the optimizer step anyway because of the lr warmup, but yes, that would definitely simplify implementation of gans and other multi-stream nets. 

My current workaround is `return nn.L1Loss(torch.zeros(1,requires_grad=True),torch.zeros(1,requires_grad=True))`

**Comment by awaelchli** [comment]
@fperazzi You can now return None from the training step, this will skip optimization, so  you can do your updates or whatever without being force to return a dummy loss. There is also a [manual optimization mode](https://pytorch-lightning.readthedocs.io/en/stable/optimizers.html#manual-optimization) now.
